[AArch64][SVE] Add patterns for bit-select instructions.

rj-jesus opened the pull request:
This patch adds patterns to select bit-sel instructions such as BSL from
(or (and a, c), (and b, (vnot c)))) and other similar patterns. For example:
```cpp
svuint64_t bsl(svuint64_t a, svuint64_t b, svuint64_t c) {
  return (a & c) | (b & ~c);
}
```
Currently:
```gas
bsl:
  and     z0.d, z2.d, z0.d
  bic     z1.d, z1.d, z2.d
  orr     z0.d, z0.d, z1.d
  ret
```
Becomes:
```gas
bsl:
  bsl     z0.d, z0.d, z1.d, z2.d
  ret
```

(EDIT: Comparison with GCC: https://godbolt.org/z/es7W5rKrY.)

rj-jesus committed:
Add tests.

rj-jesus committed:
[AArch64][SVE] Add patterns for bit-select instructions.

We are not selecting BSL/NBSL/BSL1N/BSL2N in some cases, e.g.:
```cpp
svuint64_t bsl(svuint64_t a, svuint64_t b, svuint64_t c) {
  return (a & c) | (b & ~c);
}
```

Currently generates:
```gas
bsl:
  and     z0.d, z2.d, z0.d
  bic     z1.d, z1.d, z2.d
  orr     z0.d, z0.d, z1.d
  ret
```

Instead of:
```gas
bsl:
  bsl     z0.d, z0.d, z1.d, z2.d
  ret
```

This patch adds patterns to match (or (and a, c), (and b, (vnot c)))) to
BSL, and similar derivative patterns for the other bit-sel instructions.

ktkachov commented:
Note you can use these instructions for Neon modes as well https://godbolt.org/z/ccdEoE1G7

rj-jesus commented:
> Note you can use these instructions for Neon modes as well https://godbolt.org/z/ccdEoE1G7

I was thinking we could add these cases in a subsequent PR, but I'm happy to do it here if others agree.

paulwalker-arm commented:
If agreeable I think now is the time for a bit of refactoring.  Given `AArch64ISD::BSP` exists what do you think to the idea of extending `performIntrinsicCombine` to lower `int_aarch64_sve_bsl`, `int_aarch64_sve_bsl1n`, `int_aarch64_sve_bsl2n` and `int_aarch64_sve_nbsl` to use `AArch64ISD::BSP`? That will reduce the number of patterns required for selection whilst meaning the intrinsics can benefit from any potential combines involving `AArch64ISD::BSP`.  Given at worst the intrinsics lower to `AArch64ISD::BSP` and a `NOT` I don't think we risk not being able to reconstitute the original input.

paulwalker-arm commented:
> > Note you can use these instructions for Neon modes as well https://godbolt.org/z/ccdEoE1G7
> 
> I was thinking we could add these cases in a subsequent PR, but I'm happy to do it here if others agree.

I agree it would be better for a subsequent PR to extend the fixed length support.

rj-jesus commented:
> If agreeable I think now is the time for a bit of refactoring. Given `AArch64ISD::BSP` exists, what do you think to the idea of extending `performIntrinsicCombine` to lower `int_aarch64_sve_bsl`, `int_aarch64_sve_bsl1n`, `int_aarch64_sve_bsl2n` and `int_aarch64_sve_nbsl` to use `AArch64ISD::BSP`? That will reduce the number of patterns required for selection whilst meaning the intrinsics can benefit from any potential combines involving `AArch64ISD::BSP`. Given at worst the intrinsics lower to `AArch64ISD::BSP` and a `NOT` I don't think we risk not being able to reconstitute the original input.

I think that sounds like a good idea. I suppose the other option would be to lower the intrinsics to elementary IR in `instCombineIntrinsic`, but as you say it may be too difficult to reconstitute the BSL patterns reliably later on.

Would you like me to add the lowering to `AArch64ISD::BSP` here or in a separate PR?

paulwalker-arm commented:
> Would you like me to add the lowering to `AArch64ISD::BSP` here or in a separate PR?

Up to you.  I don't mind the refactoring and the improved code generation being within this one PR.

rj-jesus committed:
Lower bit-sel intrinsics to AArch64ISD::BSP.

rj-jesus commented:
> Up to you. I don't mind the refactoring and the improved code generation being within this one PR.

Thanks, that sounds good - I've just updated the PR with the lowering to `AArch64ISD::BSP` and simplified patterns. Please let me know if this is what you had in mind or if I missed anything. :)

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:
```diff
@@ -545,12 +545,18 @@ def AArch64umulh : PatFrag<(ops node:$op1, node:$op2),
+
+def AArch64bsl2n : PatFrags<(ops node:$Op1, node:$Op2, node:$Op3),
+                            [(AArch64bsl node:$Op1, (vnot node:$Op2), node:$Op3),
+                             (or (and node:$Op1, node:$Op3), (vnot (or node:$Op2, node:$Op3)))]>;
```
That De Morgan has a lot to answer for :(

paulwalker-arm reviewed (approved):
Looks good to me.
